HttpReceiver/src/commonMain/kotlin/io/manycore/receiver/http/model/MimeMessage.kt

```diff
@@ -0,0 +1,120 @@
+package io.manycore.receiver.http.model
+
+import io.ktor.utils.io.*
+import io.ktor.utils.io.core.*
+import kotlinx.coroutines.delay
+import okio.Buffer
+import kotlin.time.Duration.Companion.milliseconds
+
+class MimeMessage private constructor(
+    private val source: ByteReadChannel,
+) {
+
+    companion object {
+
+        private val endOfHeadersDelimiters = arrayOf(
+            "\n\n",
+            "\r\n\r\n",
+        )
+
+        private val uuidRegex = Regex("[a-z0-9]{32}")
+
+        // suspending factory to be able to call parseHeaders
+        suspend operator fun invoke(source: ByteReadChannel): MimeMessage =
+            MimeMessage(source).apply { parseHeaders() }
+
+    }
+
+    /**
+     * All headers
+     */
+    lateinit var headers: MimeMessageHeaders
+        private set
+
+    /**
+     * Priority of this Message
+     */
+    val priority: MimeMessagePriority
+        get() =
+            when (val rawPriority = headers["X-Priority"]?.last()) {
+                "0"       -> MimeMessagePriority.HIGH
+                "1", null -> MimeMessagePriority.DEFAULT
+                else      -> throw IllegalArgumentException("Unexpected value '$rawPriority' for header X-Priority (expected 0 or 1)")
+            }
+
+    /**
+     * Unique ID of this Message
+     */
+    val uuid: String
+        get() {
+            val rawUuid = headers["X-Uuid"]?.last()
+            return when {
+                rawUuid == null             -> throw IllegalArgumentException("Missing header X-Uuid")
+                !rawUuid.matches(uuidRegex) -> throw IllegalArgumentException("Malformed header X-Uuid, expecting $uuidRegex")
+                else                        -> rawUuid
+            }
+        }
+
+    private val seenBytes = Buffer()
+
+    /**
+     * Reads the beginning of [source] until we read all headers, then parse them.
+     *
+     * @see headers
+     */
+    suspend fun parseHeaders() {
+        headers = MimeMessageHeaders(readRawHeaders())
+    }
+
+    /**
+     * Checks that [priority] and [uuid] headers are as expected.
+     */
+    fun verifyHeaders() {
+        priority
+        uuid
+    }
+
+    private suspend fun readRawHeaders(): String {
+        val buffer = ByteArray(512)
+        while (!source.isClosedForRead) {
+
+            // Read a bunch of bytes
+            val read = source.readAvailable(buffer)
+
+            if (read == -1) {
+                // Source is closed
+                break
+            }
+
+            if (read == 0) {
+                // Nothing available to read, wait
+                delay(10.milliseconds)
+                continue
+            }
+
+            // Save what we read for later
+            seenBytes.write(buffer, 0, read)
+
+            // Stop if what we read contain a delimiter
+            val seenBytesAsString = seenBytes.snapshot().utf8()
```

Comment:
This will fail sutbtly sometimes.

The actual email can be of any encoding or mimetype, so there will be situations where we try to cast to utf8-string where it will not work. Also, even if valid utf8() we need to be sure we're not 2 bytes into a 4 byte character etc.

I think it's better if we search for the \n\n and \r\n\r\n in a byte array, and when found cast the beginning to ascii.

Reply:
I agree, I'll do the search in bytes instead of in a String. Not taking the CPU time and RAM space to convert to String is also a bonus.